src/contracts/SavingCircles.sol

```diff
@@ -0,0 +1,276 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import {OwnableUpgradeable} from '@openzeppelin-upgradeable/access/OwnableUpgradeable.sol';
+import {IERC20} from '@openzeppelin/token/ERC20/IERC20.sol';
+import {ReentrancyGuard} from '@openzeppelin/utils/ReentrancyGuard.sol';
+
+import {ISavingCircles} from '../interfaces/ISavingCircles.sol';
+
+/**
+ * @title Saving Circles
+ * @notice Simple implementation of a rotating savings and credit association (ROSCA) for ERC20 tokens
+ * @author Breadchain Collective
+ * @author @RonTuretzky
+ * @author @bagelface
+ */
+contract SavingCircles is ISavingCircles, ReentrancyGuard, OwnableUpgradeable {
+  uint256 public constant MINIMUM_MEMBERS = 2;
+
+  mapping(bytes32 id => Circle circle) public circles;
+  mapping(address token => bool status) public allowedTokens;
+  mapping(bytes32 id => mapping(address token => uint256 balance)) public balances;
+  mapping(bytes32 id => mapping(address member => bool status)) public isMember;
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() {
+    _disableInitializers();
+  }
+
+  function initialize(address _owner) external override initializer {
+    __Ownable_init_unchained(_owner);
+  }
+
+  /**
+   * @notice Create a new saving circle
+   * @param _circle A new saving circle
+   */
+  function create(Circle memory _circle) external override {
+    bytes32 _id = keccak256(abi.encodePacked(_circle.name));
```

Comment:
I was thinking a lot about how maybe  the ID should be a hash of all the variables + timestamp? so there can be name collisions. Not sure if it's worth it or not.

Reply:
tbh i cant think of any reason to not just use serial ids (like you would do if you were just minting new NFTs) and return the value when a new circle is created

Reply:
i changed to use serial ids